onnxruntime/core/codegen/target/tvm_schedule_builder.h

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/codegen/target/tvm_scheduler.h"
+#include "core/common/common.h"
+
+namespace onnxruntime {
+namespace tvm_codegen {
+
+// TVMScheduleBuilder contains all applicable TVM scheduler passes.
+// Scheduler passes are stored in multiple dispatchers
+// that check different conditions of a tvm::Tensor.
+
+// If a tvm::Tensor satisfies more than one TVM scheduler passes,
+// the first dispatched pass will be applied.
+
+class TVMScheduleBuilder {
```

Comment:
>class TVMScheduleBuilder  [](start = 0, length = 25)

So what's the usage of this ScheduleBuilder? iterate each op in the tvm graph and apply a schedule pass? i feel in many case, the schedule pass we write is not this kind of per-op schedule.

Reply:
It is not really per-op schedule. Context can hold a graph and corresponding analysis. 
Tensor can be used as Key to get previous analyzed result and applied to schedule

Reply:
I guess your implementation won't use the default scheduler at all. This is more for sharing the logic between ARM/X86/etc

---
In reply to: [294454687](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294454687) [](ancestors = 294454687)

Reply:
then should we move the IRBuilder and schedule that only works for ARM/x86/etc to a seperate package? 


---
In reply to: [297448953](https://github.com/microsoft/onnxruntime/pull/888#discussion_r297448953) [](ancestors = 297448953,294454687)

Reply:
if that's the case, why you put tvm::Tensor as the key in the signature? I feel the way you mentioned is more like a hacky workaround to use this interface for cross-op schedule, but it will make the interface more confusing.

---
In reply to: [294551886](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294551886) [](ancestors = 294551886)

Reply:
The design separates analysis from transformation. You apply analysis, record where and what you want to transform, and then apply the transformation. tvm::Tensor is the finest unit in low IR, and Ort::Node is the finest unit in high IR. That is why we use these two as default key. You can use any other key in customized CodeGenContext. After tvm::IR is built, it is just a graph of tvm::Tensor. Op boundary is not that important.

Reply:
> then should we move the IRBuilder and schedule that only works for ARM/x86/etc to a seperate package?
> 
I had same concern and discussed with Ke a few weeks ago.